### PR TITLE
Use deterministic sort (fixes #1308)

### LINKED ISSUE #1308: Undeterministic version sort order in lockfile
**Do you want to request a _feature_ or report a _bug_?**

Bug

**What is the current behavior?**

If a transitive dependency version is specified using `X` (e.g. `3.X.X`) in multiple dependencies (and inconsistently, sometimes 'x') then lowercase and uppercase version are sorted randomly in the lockfile. This results in unwanted changes to the `yarn.lock` file (and git marking it dirty)

**If the current behavior is a bug, please provide the steps to reproduce.**

``` sh
git clone https://github.com/csabapalfi/yarn-undeterministic-sort
yarn 
rm -rf node_modules && yarn
```

Repeat the last step a few times and you should see the yarn.lock file changing even if it doesn't need to.

**What is the expected behavior?**

Depency version are always sorted the same way.

https://github.com/yarnpkg/yarn/blob/master/src/util/misc.js#L5 should work without lowercasing first but I'm not sure if any other usages of it depend on that.

**Please mention your node.js, yarn and operating system version.**
- yarn: 0.16.1
- node: v6.7.0, v6.9.0
- OSX: 10.11.6, 10.12

### PR DESCRIPTION
**Pull request by zertosh**
**Summary**

The previous sort function used `localeCompare` which is dependent on the user's locale configuration (and I think node support). This resulted in non-deterministic lock files. This new approach uses unicode point code as the rank, which is consistent for everyone.

**Test plan**

Added a test.

cc: @bestander @Daniel15

### CODE REVIEW

File: src/util/misc.js [hunk 1]
```diff
@@ -3,8 +3,23 @@
...
+  for (let i = 0; i < shortLen; i++) {
+    const aChar = a.charCodeAt(i);
+    const bChar = b.charCodeAt(i);
+    if (aChar < bChar) {
+      return -1;
+    } else if (aChar > bChar) {
+      return 1;
+    }
+  }
+  if (a.length < b.length) {
```
Comment (Daniel15) [inline src/util/misc.js]: This can just be `return b.length - a.length`

File: src/util/misc.js [hunk 2]
```diff
@@ -3,8 +3,23 @@
...
 
 export function sortAlpha(a: string, b: string): number {
-  // sort alphabetically
-  return a.toLowerCase().localeCompare(b.toLowerCase());
+  // sort alphabetically in a deterministic way
+  const shortLen = Math.min(a.length, b.length);
+  for (let i = 0; i < shortLen; i++) {
+    const aChar = a.charCodeAt(i);
+    const bChar = b.charCodeAt(i);
+    if (aChar < bChar) {
```
Comment (Daniel15) [inline src/util/misc.js]:
```
if (aChar != bChar) {
  return bChar - aChar;
}
```

**Comment by Daniel15** [comment]
The old version was case-insensitive-ish (by converting to lowercase), the new version isn't. Would that be a problem?

File: src/util/misc.js (hunk 1 above)
Comment (chicoxyzzy) [inline src/util/misc.js]: `return a.length - b.length` according to code

File: src/util/misc.js (hunk 2 above)
Comment (chicoxyzzy) [inline src/util/misc.js]: should be `return aChar - bChar;` according to code

Comment (Daniel15) [inline src/util/misc.js]: Oops, my mistake, I got the variables the wrong way around 😆

**Comment by zertosh** [comment]
> The old version was case-insensitive-ish (by converting to lowercase), the new version isn't. Would that be a problem?

I think it **should** be case sensitive. https://www.npmjs.com/package/JSONStream and https://www.npmjs.com/package/jsonstream are distinct packages.

> shouldn't the default sort() (assuming stability) be deterministic?

@kentor I'm can't figure out if that's true. MDN seems to imply that it is. Looking at [v8's implementation](https://github.com/v8/v8/blob/12af412/src/js/array.js#L715-L725), isn't very helpful either - what does `str1 < str2` do?

**Comment by zertosh** [comment]
I don't know what the right thing to do is :/

**Comment by kentor** [comment]
I'm pretty x < y just goes through every char similar to your implementation.

**Comment by Daniel15** [comment]
> I'm pretty x < y just goes through every char similar to your implementation

Hmm, I wonder if it's defined in the spec. `x < y` is pretty vague, especially in a language that uses Unicode. For example, it's not clear whether it does case folding (where characters like `å` and `ä` are considered equal to `a`). It might be implementation-specific, in which case the behaviour can change between Node.js versions and having custom code (like in this diff) is more reliable.

**Comment by zertosh** [comment]
> It might be implementation-specific, in which case the behaviour can change between Node.js versions and having custom code (like in this diff) is more reliable.

That's my inclination too. Feel free to merge at your leisure, or leave this open for more comments.

**Comment by bestander** [comment]
Let's merge this considering that it is an improvement over current sort, we can switch to implied one later

**Pull request merged**